components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -0,0 +1,159 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/master/LICENSE ).
+use crate::LocaleCanonicalizer;
+use crate::LocaleCanonicalizerError;
+use icu_locid::subtags;
+use icu_locid::LanguageIdentifier;
+use icu_locid::Locale;
+use icu_provider::prelude::*;
+use icu_provider::structs::likelysubtags::*;
+use std::borrow::Cow;
+
+impl LocaleCanonicalizer<'_> {
+    /// A constructor which takes a DataProvider and creates a
+    /// LocaleCanonicalizer.
+    pub fn new<'a>(
+        provider: &dyn DataProvider<'a, LikelySubtagsV1>,
+    ) -> Result<LocaleCanonicalizer<'a>, DataError> {
+        let payload: Cow<LikelySubtagsV1> = provider
+            .load_payload(&DataRequest::from(key::LIKELY_SUBTAGS_V1))?
+            .take_payload()?;
+
+        Ok(LocaleCanonicalizer {
+            likely_subtags: payload,
+        })
+    }
+
+    /// This returns a new Locale that is the result of running the
+    /// 'Add Likely Subtags' algorithm from
+    /// https://www.unicode.org/reports/tr35/#Likely_Subtags.
+    pub fn maximize(&self, locale: &mut Locale) -> Result<bool, LocaleCanonicalizerError> {
```

Comment:
I prefer avoiding bools in public APIs, except in situations where the answer is truly binary.  Here, I can think of several possible return values; for example:

- The locale was found in data and modified.
- The locale was found in data but was already maximized.
- The locale was not found in data and not modified.

Maybe you could even go further and return which fields were changed (language, script, region).  In any case, I suggest changing the return value to an enum rather than a bool.